src/effects/Compressor.cpp

```diff
@@ -1,737 +1,96 @@
-      // to avoid accumulation of rounding errors
-      FreshenCircle();
-   }
-   // First apply a peak detect with the requested decay rate
-   last = mLastLevel;
-   for(size_t i=0; i<len; i++) {
-      if(mUsePeak)
-         level = fabs(buffer[i]);
-      else // use RMS
-         level = AvgCircle(buffer[i]);
-      // Don't increase gain when signal is continuously below the noise floor
-      if(level < mNoiseFloor) {
-         mNoiseCounter++;
-      } else {
-         mNoiseCounter = 0;
-      }
-      if(mNoiseCounter < 100) {
-         last *= mDecayFactor;
-         if(last < mThreshold)
-            last = mThreshold;
-         if(level > last)
-            last = level;
-      }
-      env[i] = last;
-   }
-   mLastLevel = last;
-
-   // Next do the same process in reverse direction to get the requested attack rate
-   last = mLastLevel;
-   for(size_t i = len; i--;) {
-      last *= mAttackInverseFactor;
-      if(last < mThreshold)
-         last = mThreshold;
-      if(env[i] < last)
-         env[i] = last;
-      else
-         last = env[i];
-   }
-
-   if((previous != NULL) && (previous_len > 0)) {
-      // If the previous envelope was passed, propagate the rise back until we intersect
-      for(size_t i = previous_len; i--;) {
-         last *= mAttackInverseFactor;
-         if(last < mThreshold)
-            last = mThreshold;
-         if(previous[i] < last)
-            previous[i] = last;
-         else // Intersected the previous envelope buffer, so we are finished
-            return;
-      }
-      // If we can't back up far enough, project the starting level forward
-      // until we intersect the desired envelope
-      last = previous[0];
-      for(size_t i=1; i<previous_len; i++) {
-         last *= mAttackFactor;
-         if(previous[i] > last)
-            previous[i] = last;
-         else // Intersected the desired envelope, so we are finished
-            return;
-      }
-      // If we still didn't intersect, then continue ramp up into current buffer
-      for(size_t i=0; i<len; i++) {
-         last *= mAttackFactor;
-         if(buffer[i] > last)
-            buffer[i] = last;
-         else // Finally got an intersect
-            return;
-      }
-      // If we still didn't intersect, then reset mLastLevel
-      mLastLevel = last;
-   }
-}
-
-float EffectCompressor::DoCompression(float value, double env)
+auto EffectCompressor::RealtimeSupport() const -> RealtimeSince
 {
-   float out;
-   if(mUsePeak) {
-      // Peak values map 1.0 to 1.0 - 'upward' compression
-      out = value * pow(1.0/env, mCompression);
-   } else {
-      // With RMS-based compression don't change values below mThreshold - 'downward' compression
-      out = value * pow(mThreshold/env, mCompression);
-   }
-
-   // Retain the maximum value for use in the normalization pass
-   if(mMax < fabs(out))
-      mMax = fabs(out);
-
-   return out;
+   return RealtimeSince::After_3_1;
```

Comment:
I'm not sure here. Eventually it will replace the offline processor, "stealing" its name (currently "Real-time Compressor"), so maybe it should stay `After_3_1` and not be `Always` ? On the other hand, I don't see any different consequence in returning `After_3_1` or `Always`.

Reply:
If you reuse the old ID, you should introduce After_3_6 (3_7?)

This is a way Paul has introduced to distinguish between RT/non-RT versions of built in effects